src/lib/message/L2ToL1Message.ts

```diff
@@ -376,82 +365,88 @@ export class L2ToL1MessageReader extends L2ToL1Message {
   }
 
   /**
-   * Estimates the L1 block number in which this L2 to L1 tx will be available for execution
+   * Estimates the L1 block number in which this L2 to L1 tx will be available for execution.
+   * If the message can or already has been executed, this returns null
    * @param l2Provider
-   * @returns expected L1 block number where the L2 to L1 message will be executable
+   * @returns expected L1 block number where the L2 to L1 message will be executable. Returns null if the message can or already has been executed
    */
   public async getFirstExecutableBlock(
     l2Provider: Provider
-  ): Promise<BigNumber> {
-    throw new ArbTsError('getFirstExecutableBlock not implemented')
-    /*
-    // expected number of L1 blocks that it takes for an L2 tx to be included in a L1 assertion
-    const ASSERTION_CREATED_PADDING = 50
-    // expected number of L1 blocks that it takes for a validator to confirm an L1 block after the node deadline is passed
-    const ASSERTION_CONFIRMED_PADDING = 20
-
+  ): Promise<BigNumber | null> {
     // TODO: create version that queries multiple L2 to L1 txs, so a single multicall can make all requests
     // we assume the L2 to L1 tx is valid, but we could check that on the constructor that the L2 to L1 msg is valid
-    const network = await getL2Network(l2Provider)
+    const l2Network = await getL2Network(l2Provider)
 
-    // TODO: use IRollupUser interface instead
-    const rollup = RollupUserFacet__factory.connect(
-      network.ethBridge.rollup,
+    const rollup = RollupUserLogic__factory.connect(
+      l2Network.ethBridge.rollup,
       this.l1Provider
     )
 
-    const proof = await this.tryGetProof(l2Provider)
-    // here we assume the L2 to L1 tx is actually valid, so the user needs to wait the max time.
-    if (proof === null)
-      return BigNumber.from(network.confirmPeriodBlocks)
-        .add(ASSERTION_CREATED_PADDING)
-        .add(ASSERTION_CONFIRMED_PADDING)
-    // we can't check if the L2 to L1 tx isSpent on the outbox, so we instead try executing it
-    if (await this.hasExecuted(proof)) return BigNumber.from(0)
+    const status = await this.status(l2Provider)
+    if (status === L2ToL1MessageStatus.EXECUTED) return null
+    if (status === L2ToL1MessageStatus.CONFIRMED) return null
+    if (status === L2ToL1MessageStatus.NOT_FOUND)
+      throw new ArbTsError('L2ToL1Msg not found')
+
+    // consistency check in case we change the enum in the future
+    if (status !== L2ToL1MessageStatus.UNCONFIRMED)
+      throw new ArbTsError('L2ToL1Msg expected to be unconfirmed')
+
     const latestBlock = await this.l1Provider.getBlockNumber()
 
     const eventFetcher = new EventFetcher(this.l1Provider)
 
-    const events = (
+    const logs = (
```

Comment:
thinking a logs.length > 0 sanity check (to throw a nice error in the Bad Situation where nodes haven't been created for a long while)

Reply:
added in the case